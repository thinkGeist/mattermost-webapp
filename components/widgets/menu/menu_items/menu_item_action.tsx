// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import classNames from 'classnames';

import OverlayTrigger from 'components/overlay_trigger';
import Tooltip from 'components/tooltip';

import Constants from 'utils/constants';

import menuItem from './menu_item';

type Props = {
    onClick: (e: React.MouseEvent) => void;
    ariaLabel?: string;
    text: React.ReactNode;
    extraText?: string;
    id?: string;
    buttonClass?: string;
    rightDecorator?: React.ReactNode;
    isDangerous?: boolean;
    disabled?: boolean;
    disabledText?: string;
}

export const MenuItemActionImpl = ({
    onClick,
    ariaLabel,
    text,
    extraText,
    id,
    buttonClass,
    rightDecorator,
    isDangerous,
    disabled,
    disabledText,
}: Props) => {
    const disabledTooltip = (
        <Tooltip id='disabled-tooltip'>
            {disabledText}
        </Tooltip>
    );
    return (
        <OverlayTrigger
            disabled={!disabled}
            placement='left'
            delayShow={Constants.OVERLAY_TIME_DELAY}
            trigger={Constants.OVERLAY_DEFAULT_TRIGGER}
            overlay={disabledTooltip}
        >
            <button
                data-testid={id}
                id={id}
                aria-label={ariaLabel}
                className={classNames('style--none', buttonClass, {
                    'MenuItem__with-help': extraText,
                    MenuItem__dangerous: isDangerous,
                    disabled,
                })}
                onClick={onClick}
                disabled={disabled}
            >
                {text && <span className='MenuItem__primary-text'>{text}{rightDecorator}</span>}
                {extraText && <span className='MenuItem__help-text'>{extraText}</span>}
            </button>
        </OverlayTrigger>
    );
};

const MenuItemAction = menuItem(MenuItemActionImpl);
MenuItemAction.displayName = 'MenuItemAction';

export default MenuItemAction;
