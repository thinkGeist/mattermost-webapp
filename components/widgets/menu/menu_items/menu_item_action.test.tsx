// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {MenuItemActionImpl} from './menu_item_action';

describe('components/MenuItemAction', () => {
    test('should match snapshot', () => {
        const wrapper = shallow(
            <MenuItemActionImpl
                onClick={jest.fn()}
                text='Whatever'
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });
    test('should match snapshot with extra text', () => {
        const wrapper = shallow(
            <MenuItemActionImpl
                onClick={jest.fn()}
                text='Whatever'
                extraText='Extra Text'
            />,
        );

        expect(wrapper).toMatchSnapshot();
    });
});
