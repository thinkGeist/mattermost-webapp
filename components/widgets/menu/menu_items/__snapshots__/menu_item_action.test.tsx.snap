// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MenuItemAction should match snapshot 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  delayShow={400}
  disabled={true}
  overlay={
    <Tooltip
      id="disabled-tooltip"
    />
  }
  placement="left"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <button
    className="style--none"
    onClick={[MockFunction]}
  >
    <span
      className="MenuItem__primary-text"
    >
      Whatever
    </span>
  </button>
</OverlayTrigger>
`;

exports[`components/MenuItemAction should match snapshot with extra text 1`] = `
<OverlayTrigger
  defaultOverlayShown={false}
  delayShow={400}
  disabled={true}
  overlay={
    <Tooltip
      id="disabled-tooltip"
    />
  }
  placement="left"
  trigger={
    Array [
      "hover",
      "focus",
    ]
  }
>
  <button
    className="style--none MenuItem__with-help"
    onClick={[MockFunction]}
  >
    <span
      className="MenuItem__primary-text"
    >
      Whatever
    </span>
    <span
      className="MenuItem__help-text"
    >
      Extra Text
    </span>
  </button>
</OverlayTrigger>
`;
